REGRESSION (293770@main): clicking on the "+" button in the Sources tab sidebar does nothing

the-chenergy opened the pull request:
#### 41618b5f418cbaefe058789f6e99d60a7939a4b1
<pre>
REGRESSION (293770@main): clicking on the &quot;+&quot; button in the Sources tab sidebar does nothing
<a href="https://rdar.apple.com/153193833">rdar://153193833</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=294392">https://bugs.webkit.org/show_bug.cgi?id=294392</a>

Reviewed by Patrick Angle.

If there are obscuredContentInsets applied to the WKInspectorWKWebView,
when InspectorFrontendHost is commanded to show a context menu, it
shouldn&apos;t base its location entirely on the MouseEvent that came from
the inspector&apos;s frontend. The frontend is unaware of the obscuredContentInsets
while the frame is. InspectorFrontendHost needs to apply that offset,
as otherwise the mouse event that the frame later creates to bring up
the context menu might happen on top of a wrong event target, one that
doesn&apos;t have a contextmenu event handler which would insert the expected
menu&apos;s contents.

Tested manually on macOS 26.0 where this bug also reproduced, while the
inspector having either a left inset (with the browser&apos;s sidebar open)
or a top inset (with the inspector undocked).

* Source/WebCore/inspector/InspectorFrontendHost.cpp:
(WebCore::InspectorFrontendHost::dispatchEventAsContextMenuEvent):

Canonical link: <a href="https://commits.webkit.org/296458@main">https://commits.webkit.org/296458@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8796a7feeca67bcca51a1f6462fe9012a0469229

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108603 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28264 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18688 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113811 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/58993 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110566 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28953 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36818 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/82490 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/58993 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111551 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/22982 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/97826 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/62927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/22400 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/15964 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/58525 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/92351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/16015 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116932 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35657 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/26291 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/91513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36030 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/94096 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/91317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/36216 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/13979 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17535 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35558 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/41091 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35268 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38614 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36946 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/inspector/InspectorFrontendHost.cpp:
```diff
@@ -611,8 +611,13 @@ void InspectorFrontendHost::dispatchEventAsContextMenuEvent(Event& event)
     auto& mouseEvent = downcast<MouseEvent>(event);
-    auto& frame = *downcast<Node>(mouseEvent.target())->document().frame();
-    m_frontendPage->contextMenuController().showContextMenuAt(frame, roundedIntPoint(mouseEvent.absoluteLocation()));
+    LocalFrame& frame = *downcast<Node>(mouseEvent.target())->document().frame();
+    LayoutPoint location = mouseEvent.absoluteLocation();
+    if (RefPtr<LocalFrameView> view = frame.view()) {
```
```suggestion
    auto& frame = *downcast<Node>(mouseEvent.target())->document().frame();
    auto location = mouseEvent.absoluteLocation();
    if (RefPtr view = frame.view()) {
```

patrickangle commented on Source/WebCore/inspector/InspectorFrontendHost.cpp:
```diff
@@ -611,8 +611,13 @@ void InspectorFrontendHost::dispatchEventAsContextMenuEvent(Event& event)
+        FloatBoxExtent insets = view->obscuredContentInsets();
+        location.move(insets.left(), insets.top());
+    }
+    m_frontendPage->contextMenuController().showContextMenuAt(frame, roundedIntPoint(location));
```
nit: newline should separate the end of the block above from this line of code

patrickangle reviewed (approved):
The other place I'd suggest confirming functionality is Element Selection Mode, where you click on the reticle icon in Web Inspector and then can hover over elements to view info about them and click on one to select it. Please confirm that content is all still aligned correctly and that clicks aren't offset in that mode either.

the-chenergy committed:
REGRESSION (293770@main): clicking on the "+" button in the Sources tab sidebar does nothing
rdar://153193833
https://bugs.webkit.org/show_bug.cgi?id=294392

Reviewed by Patrick Angle.

If there are obscuredContentInsets applied to the WKInspectorWKWebView,
when InspectorFrontendHost is commanded to show a context menu, it
shouldn't base its location entirely on the MouseEvent that came from
the inspector's frontend. The frontend is unaware of the obscuredContentInsets
while the frame is. InspectorFrontendHost needs to apply that offset,
as otherwise the mouse event that the frame later creates to bring up
the context menu might happen on top of a wrong event target, one that
doesn't have a contextmenu event handler which would insert the expected
menu's contents.

Tested manually on macOS 26.0 where this bug also reproduced, while the
inspector having either a left inset (with the browser's sidebar open)
or a top inset (with the inspector undocked).

* Source/WebCore/inspector/InspectorFrontendHost.cpp:
(WebCore::InspectorFrontendHost::dispatchEventAsContextMenuEvent):

Canonical link: https://commits.webkit.org/296458@main
